storage/gcs/gcs.go

```diff
@@ -399,6 +399,86 @@ func (s *Storage) Sequence(ctx context.Context, leaf []byte) (uint64, error) {
 	return s.pool.Add(leaf)
 }
 
+// NextAvailable returns the next available unassigned index.
+func (s *Storage) NextAvailable(ctx context.Context) (uint64, error) {
+	tx, err := s.dbPool.BeginTx(ctx, nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to begin tx: %v", err)
+	}
+	defer func() {
+		if tx != nil {
+			tx.Rollback()
+		}
+	}()
+
+	r := tx.QueryRowContext(ctx, "SELECT id, next FROM SeqCoord WHERE id = ? FOR UPDATE", 0)
+
+	var id, next uint64
+	if err := r.Scan(&id, &next); err == sql.ErrNoRows {
+		return 0, fmt.Errorf("init new log in seqcoord: %v", err)
+	} else if err != nil {
+		return 0, fmt.Errorf("failed to read seqcoord: %v", err)
+	}
+
+	return next, nil
+}
+
+// AddSequenced commits leaves to the log starting at the index of startSeq.
+func (s *Storage) AddSequenced(ctx context.Context, startSeq uint64, leaves [][]byte) error {
+	batch := writer.Batch{
+		Entries: leaves,
+	}
+
+	b := &bytes.Buffer{}
+	e := gob.NewEncoder(b)
+	if err := e.Encode(batch); err != nil {
+		return fmt.Errorf("failed to serialise batch: %v", err)
+	}
+	data := b.Bytes()
+	num := len(batch.Entries)
+
+	tx, err := s.dbPool.BeginTx(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("failed to begin tx: %v", err)
+	}
+	defer func() {
+		if tx != nil {
+			tx.Rollback()
+		}
+	}()
+
+	// Check that the start index of the leaves provided is consistent with the
+	// next unassigned sequence number.
+	r := tx.QueryRowContext(ctx, "SELECT id, next FROM SeqCoord WHERE id = ? FOR UPDATE", 0)
+	var id, next uint64
+	if err := r.Scan(&id, &next); err == sql.ErrNoRows {
+		klog.Info("New log - first sequence")
+		if _, err := tx.ExecContext(ctx, "INSERT INTO SeqCoord (id, next) VALUES (?, ?)", 0, next+uint64(num)); err != nil {
```

Comment:
This seems like a bug - presumably the log is empty at this point (which is why there's no seqcoord row), but we write in `0+num` and then return without allowing the code below to actually store `data` into `Seq` at `0`.

BTW, in the newer Spanner impl, I've removed need for this check here by just having `initDB` attempt to insert the default row into both `SeqCoord` and `IntCoord` - could do that here too?

Reply:
That's cleaner, thanks!